### PR TITLE
Configurable record processor

### PR DESCRIPTION
**Pull request by jmatusewicz**
This PR enables checkpointing of single records and infinite retries of record processing. 

Until now checkpointing happens only after a whole batch is processed. This may lead into undesirable  situations where one record will be processed multiple times:
Scenario: Application is shutdown while processing a batch of n-records, but only a subset of records of the batch was processed, so no checkpoint was saved. On next application start the whole batch would be processed again, hence the first records of this batch would be processed twice.

This PR allows to set the checkpoint strategy to be either `BATCH` (default and old behaviour) or `RECORD` (new behaviour).
Furthermore, you can configure the retry attempts to be infinitely. This in combination with `RECORD` checkpointing strategy would make sure that no record would be ever skipped and every record would be processed only a single time.

### CODE REVIEW

File: src/integrationTest/kotlin/de/bringmeister/spring/aws/kinesis/CheckpointingIntegrationTest.kt [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+                .build()
+
+            return spy(KinesisClientLeaseManager(STREAM, dynamoDbClient))
+        }
+    }
+
+    @Autowired
+    lateinit var outbound: AwsKinesisOutboundGateway
+
+    @Autowired
```
Comment (jrehwaldt) [inline src/integrationTest/kotlin/de/bringmeister/spring/aws/kinesis/CheckpointingIntegrationTest.kt]: What about `@SpyBean`?

File: src/integrationTest/kotlin/de/bringmeister/spring/aws/kinesis/CheckpointingIntegrationTest.kt [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        outbound.send(STREAM, *records)
+
+        doCallRealMethod()
+            .doThrow(KinesisClientLibIOException::class.java)
+            .doCallRealMethod()
+            .`when`(spyLeaseManager).updateLease(any())
+
+        val recordsProcessed = waitForRecordsToBeProcessed()
+
+        assertThat(recordsProcessed).isTrue()
```
Comment (jrehwaldt) [inline src/integrationTest/kotlin/de/bringmeister/spring/aws/kinesis/CheckpointingIntegrationTest.kt]: Do you need to verify the calling of the checkpoint method? How do you know it did retry?
(and that it didn't in the following test)

File: src/integrationTest/kotlin/de/bringmeister/spring/aws/kinesis/CheckpointingIntegrationTest.kt [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    @Test
+    fun `should not retry checkpoiting on non-retryable exceptions`() {
```
Comment (jrehwaldt) [inline src/integrationTest/kotlin/de/bringmeister/spring/aws/kinesis/CheckpointingIntegrationTest.kt]: `checkpoiting` -> `checkpointing`

File: src/integrationTest/kotlin/de/bringmeister/spring/aws/kinesis/handler/retry/FiniteRetryRecordProcessorIntegrationTest.kt [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+import org.springframework.context.annotation.Bean
+import org.springframework.context.annotation.Import
+import org.springframework.test.context.TestPropertySource
+import org.springframework.test.context.junit4.SpringRunner
+import java.util.concurrent.TimeUnit
+
+@RunWith(SpringRunner::class)
+@ContainerTest
+@Import(FiniteRetryRecordProcessorIntegrationTest.Config::class)
+@TestPropertySource(properties = ["aws.kinesis.handler.retry.maxRetries=$MAX_RETRIES"])
```
Comment (jrehwaldt) [inline src/integrationTest/kotlin/de/bringmeister/spring/aws/kinesis/handler/retry/FiniteRetryRecordProcessorIntegrationTest.kt]: Clever 👍

File: src/main/kotlin/de/bringmeister/spring/aws/kinesis/AwsKinesisRecordProcessor.kt [hunk 5]
```diff
@@ -13,7 +13,6 @@ import com.amazonaws.services.kinesis.clientlibrary.types.ShutdownInput
 import org.slf4j.LoggerFactory
 import org.springframework.context.ApplicationEventPublisher
 import com.amazonaws.services.kinesis.model.Record as AwsRecord
-import de.bringmeister.spring.aws.kinesis.Record as BmRecord
 
 class AwsKinesisRecordProcessor<D, M>(
```
Comment (jrehwaldt) [inline src/main/kotlin/de/bringmeister/spring/aws/kinesis/AwsKinesisRecordProcessor.kt]: Nice and much cleaner. Also 👍 for the decoupling of checkpoint/handler retry config.

File: src/main/kotlin/de/bringmeister/spring/aws/kinesis/retry/RetryableRecordHandlerAutoConfiguration.kt [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    @Bean
+    fun retryableRecordPostProcessor(settings: RetryableRecordProcessorSettings) = RetryableRecordHandlerPostProcessor(settings)
+}
+
+@Validated
+@ConfigurationProperties(prefix = "aws.kinesis.handler.retry")
+class RetryableRecordProcessorSettings {
+
+    companion object {
+        const val NO_RETRIES = 0
```
Comment (jrehwaldt) [inline src/main/kotlin/de/bringmeister/spring/aws/kinesis/retry/RetryableRecordHandlerAutoConfiguration.kt]: Duplicated with value from `RetrySettings`.

File: src/integrationTest/kotlin/de/bringmeister/spring/aws/kinesis/CheckpointingIntegrationTest.kt (hunk 1 above)
Comment (jmatusewicz) [reply src/integrationTest/kotlin/de/bringmeister/spring/aws/kinesis/CheckpointingIntegrationTest.kt]: I removed the test entirely as it was a kinda hack.

**Pull request merged**